pkg/i2gw/providers/istio/resource_reader.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package istio
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type reader struct {
+	k8sClient client.Client
+}
+
+func newResourceReader(k8sClient client.Client) reader {
+	return reader{
+		k8sClient: k8sClient,
+	}
+}
+
+func (r *reader) readResourcesFromCluster(ctx context.Context) (*storage, error) {
+	res := newResourcesStorage()
+
+	gateways, err := r.readGatewaysFromCluster(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read gateways: %w", err)
+	}
+
+	res.Gateways = gateways
+
+	virtualServices, err := r.readVirtualServicesFromCluster(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read virtual services: %w", err)
+	}
+
+	res.VirtualServices = virtualServices
+
+	return &res, nil
+}
+
+func (r *reader) readUnstructuredObjects(objects []*unstructured.Unstructured) (*storage, error) {
+	res := newResourcesStorage()
+
+	for _, obj := range objects {
+		if obj.GetAPIVersion() != APIVersion {
+			return nil, fmt.Errorf("got unexpected API version, want: %v, got: %v", APIVersion, obj.GetAPIVersion())
+		}
+
+		switch objKind := obj.GetKind(); objKind {
+		case GatewayKind:
+			var gw gateway
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.UnstructuredContent(), &gw); err != nil {
+				return nil, fmt.Errorf("failed to parse istio gateway object: %w", err)
+			}
+			res.Gateways[types.NamespacedName{
+				Namespace: gw.Namespace,
+				Name:      gw.Name,
+			}] = &gw
+
+		case VirtualServiceKind:
+			var vs virtualService
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.UnstructuredContent(), &vs); err != nil {
+				return nil, fmt.Errorf("failed to parse istio virtual service object: %w", err)
+			}
+
+			res.VirtualServices[types.NamespacedName{
+				Namespace: vs.Namespace,
+				Name:      vs.Name,
+			}] = &vs
+		default:
+			return nil, fmt.Errorf("unsupported resource kind: %v", objKind)
```

Comment:
Is there a way where we can warn the user and skip the resource instead of returning an error? I assume some users will have other kinds inside the same file

Reply:
added log message for both APIVersion check and Kind